File: tests/deepflow-typescript/src/services/safe.service.ts
/**
 * Safe service - Demonstrates sanitized patterns.
 *
 * These methods use secure coding practices that should be
 * recognized by TheAuditor as NOT vulnerable.
 */

import { SafeRepository } from '../repositories/safe.repository';
import { escapeHtml } from '../utils/string.utils';

export class SafeService {
  private repository: SafeRepository;

  constructor() {
    this.repository = new SafeRepository();
  }

  /**
   * Safe search using parameterized queries.
   *
   * @param query - Search query (will be parameterized, not concatenated)
   */
  async safeSearch(query: string): Promise<any> {
    return this.repository.searchSafe(query);
  }

  /**
   * Get user by ID (already validated).
   *
   * @param userId - User ID (numeric, already validated)
   */
  async getUserById(userId: string): Promise<any> {
    return this.repository.findByIdSafe(userId);
  }

  /**
   * Safe HTML rendering with escaping.
   *
   * @param content - Content to render (will be HTML escaped)
   */
  async safeRender(content: string): Promise<string> {
    // SANITIZER: HTML escape before rendering
    const safeContent = escapeHtml(content);
    return `<div class="content">${safeContent}</div>`;
  }
}

File: tests/deepflow-python/app/services/safe_service.py
"""Safe service - Demonstrates sanitized patterns.

These methods use secure coding practices that should be
recognized by TheAuditor as NOT vulnerable.
"""

from app.repositories.safe_repository import SafeRepository
from app.core.template_renderer import TemplateRenderer


class SafeService:
    """Service with safe implementations.

    All methods here use proper sanitization and should NOT
    be flagged as vulnerable.
    """

    def __init__(self):
        self.repo = SafeRepository()
        self.renderer = TemplateRenderer()

    async def get_by_email(self, email: str) -> dict:
        """Get user by email.

        The email has already been validated by regex in the route.
        This uses parameterized queries for additional safety.

        Args:
            email: SANITIZED email (regex validated)
        """
        return self.repo.find_by_email_safe(email)

    async def safe_search(self, query: str) -> dict:
        """Safe search using parameterized queries.

        Args:
            query: Search query (will be parameterized, not concatenated)
        """
        return self.repo.search_safe(query)

    async def safe_render(self, title: str) -> dict:
        """Safe template rendering with HTML escaping.

        Args:
            title: Title to render (will be HTML escaped)
        """
        html = self.renderer.render_safe(title, {"type": "safe_demo"})
        return {"html": html, "sanitized": True}
